zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbRpaState.java

```diff
@@ -0,0 +1,285 @@
+  }
+
+  @Override
+  public Optional<PersistedRpa> findLatestRpaById(
+      final DirectBuffer rpaId, final String tenantId) {
+    tenantIdKey.wrapString(tenantId);
+    final Optional<PersistedRpa> cachedRpa = getRpaFromCache(tenantId, rpaId);
+    if (cachedRpa.isPresent()) {
+      return cachedRpa;
+    }
+
+    final PersistedRpa persistedRpa = getPersistedRpaById(rpaId, tenantId);
+    if (persistedRpa == null) {
+      return Optional.empty();
+    }
+    rpasByTenantIdAndIdCache.put(new DbRpaState.TenantIdAndRpaId(tenantId, rpaId), persistedRpa);
+    return Optional.of(persistedRpa);
+  }
+
+  @Override
+  public Optional<PersistedRpa> findRpaByKey(final long rpaKey, final String tenantId) {
+    tenantIdKey.wrapString(tenantId);
+    dbRpaKey.wrapLong(rpaKey);
+    return Optional.ofNullable(rpasByKey.get(tenantAwareRpaKey)).map(PersistedRpa::copy);
+  }
+
+  @Override
+  public Optional<PersistedRpa> findRpaByIdAndDeploymentKey(
+      final DirectBuffer rpaId, final long deploymentKey, final String tenantId) {
+    tenantIdKey.wrapString(tenantId);
+    dbRpaId.wrapBuffer(rpaId);
+    dbDeploymentKey.wrapLong(deploymentKey);
+    return Optional.ofNullable(
+            rpaKeyByRpaIdAndDeploymentKeyColumnFamily.get(tenantAwareRpaIdAndDeploymentKey))
+        .flatMap(key -> findRpaByKey(key.inner().wrappedKey().getValue(), tenantId));
+  }
+
+  @Override
+  public Optional<PersistedRpa> findRpaByIdAndVersionTag(
+      final DirectBuffer rpaId, final String versionTag, final String tenantId) {
+    tenantIdKey.wrapString(tenantId);
+    dbRpaId.wrapBuffer(rpaId);
+    dbVersionTag.wrapString(versionTag);
+    return Optional.ofNullable(
+            rpaKeyByRpaIdAndVersionTagColumnFamily.get(tenantAwareRpaIdAndVersionTagKey))
+        .flatMap(key -> findRpaByKey(key.inner().wrappedKey().getValue(), tenantId));
+  }
+
+  @Override
+  public int getNextRpaVersion(final String rpaId, final String tenantId) {
+    return (int) versionManager.getHighestResourceVersion(rpaId, tenantId) + 1;
+  }
+
+  @Override
+  public void clearCache() {
+    rpasByTenantIdAndIdCache.invalidateAll();
+    versionManager.clear();
+  }
+
+  private PersistedRpa getPersistedRpaById(final DirectBuffer rpaId, final String tenantId) {
+    dbRpaId.wrapBuffer(rpaId);
+    final long latestVersion = versionManager.getLatestResourceVersion(rpaId, tenantId);
+    rpaVersion.wrapLong(latestVersion);
+    final PersistedRpa persistedRpa =
+        rpaByIdAndVersionColumnFamily.get(tenantAwareIdAndVersionKey);
+    if (persistedRpa == null) {
+      return null;
+    }
+    return persistedRpa.copy();
+  }
+
+  private Optional<PersistedRpa> getRpaFromCache(
+      final String tenantId, final DirectBuffer rpaId) {
+    return Optional.ofNullable(
+        rpasByTenantIdAndIdCache.getIfPresent(new DbRpaState.TenantIdAndRpaId(tenantId, rpaId)));
+  }
+
+  private record TenantIdAndRpaId(String tenantId, DirectBuffer rpaId) {}
```

Comment:
🔧 Better to generalize this into:

```suggestion
  private record TenantIdAndResourceId(String tenantId, DirectBuffer resourceId) {}
```